exasol/nb_connector/connections.py

```diff
@@ -60,6 +61,86 @@ def _extract_ssl_options(conf: Secrets) -> dict:
     return sslopt
 
 
+def infer_backend(conf: Secrets,
+                  backend_requirements: dict[StorageBackend, list[list[CKey]]]) -> StorageBackend:
+    """
+    Infers the backend looking at the provided configuration data. Raises a ValueError exception
+    if the configuration is insufficient for any of the backends.
+
+    Parameters:
+        conf:
+            The secret store.
+        backend_requirements:
+            Configuration requirements for each backend. The requirements should
+            be given as a list of groups of configuration element. The first group
+            contains elements all of which are required for accessing the backend.
+            Additionally, required is at least one element from each of the remaining
+            groups.
+    """
+
+    def group_match(group_no: int, key_group: list[CKey]) -> bool:
+        bool_func = all if group_no == 0 else any
+        return bool_func(conf.get(key) for key in key_group)
+
+    def all_groups_match(key_groups: list[list[CKey]]) -> bool:
+        return all(group_match(group_no, key_group)
+                   for group_no, key_group in enumerate(key_groups))
+
+    def key_names(group_no: int, key_group: list[CKey]) -> str:
+        sep = ', ' if group_no == 0 else ' or '
+        return sep.join(key.name for key in key_group)
+
+    def expected_keys(key_groups: list[list[CKey]]) -> str:
+        return ', '.join(key_names(group_no, key_group)
+                         for group_no, key_group in enumerate(key_groups))
+
+    for backend, backend_key_groups in backend_requirements.items():
+        if all_groups_match(backend_key_groups):
+            return backend
+
+    message = ('Incomplete parameter list. Please provide the parameters ' +
+               ' or '.join(f'[{expected_keys(backend_key_groups)}] for {backend.name} database'
+                           for backend, backend_key_groups in backend_requirements.items()))
+    raise ValueError(message)
+
+
+def infer_db_backend(conf: Secrets) -> StorageBackend:
+    """
+    Infers the backend for a database connection.
+    """
+    return infer_backend(
+        conf,
+        {
+            StorageBackend.onprem: [
+                [CKey.db_host_name, CKey.db_port, CKey.db_user, CKey.db_password]
+            ],
+            StorageBackend.saas: [
+                [CKey.saas_url, CKey.saas_token, CKey.saas_account_id],
+                [CKey.saas_database_id, CKey.saas_database_name]
```

Comment:
How about adding an explicit signal to the values?
E.g. 
```
{"all": [CKey.saas_url, CKey.saas_token, CKey.saas_account_id]},
{"any": [CKey.saas_database_id, CKey.saas_database_name]},
```

In this example I used dicts, but it could also be simple lists with the first element being a simple string with the semantics to trigger the bool-func.

Reply:
I think it will complicate both the data structure and the explanation of what it should be. It is already more generic than it is practically needed. Currently, I've got at most one group of options. I'd rather leave it as it is.

Reply:
:+1: 
> already more generic than it is practically needed

probably, yes :slightly_smiling_face: